R/peaks-functions.R

```diff
@@ -286,11 +310,150 @@ NULL
+        ## Keep only matches here
+        nona <- !(is.na(pmap[[1L]]) | is.na(pmap[[2L]]))
+        if (any(nona)) {
+            map[[1L]] <- c(map[[1L]], pmap[[1L]][nona])
+            map[[2L]] <- c(map[[2L]], pmap[[2L]][nona])
+            idx <- order(map[[1L]])
+            map[[1L]] <- map[[1L]][idx]
+            map[[2L]] <- map[[2L]][idx]
+        }
+    }
+    list(x = x[map[[1L]], , drop = FALSE], y = y[map[[2L]], , drop = FALSE])
+}
+
+#' @title GNPS spectra similarity score
+#'
+#' @description
+#'
+#' The GNPS approach considers in its similarity score, in addition to directly
+#' matching peaks between two spectra, also matches of peaks which difference in
+#' m/z values matches the difference of the spectra's precursor m/z. For peaks
+#' that match multiple peaks in the other spectrum the matching peak pair with
+#' the higher value/similarity is considered in the final similarity score.
+#'
+#' To calculate GNPS scores, the [joinPeaksGnps()] function should be used first
+#' to match the peaks between the compared spectra and `gnps` should then be
+#' called on the resulting aligned peak matrices. To use the GNPS score on a
+#' [Spectra()] object `compareSpectra` should be used with
+#' `MAPFUN = joinPeaksGnps` and `FUN = gnps`.
+#'
+#' @details
+#'
+#' The implementation bases on the R code from the publication listed in the
+#' references.
+#'
+#' @param x `numeric` peak `matrix` with m/z (first column) and intensity
+#'     values (second column). Rows in `x` and `y` are considered to contain
+#'     matching peaks, i.e. the first row in `x` contains a peak matching the
+#'     first peak (row) of `y` or `NA_real_` if the peak does not match.
+#'
+#' @param y `numeric` peak `matrix` with m/z (first column) and intensity
+#'     values (second column). Rows in `x` and `y` are considered to contain
+#'     matching peaks, i.e. the first row in `x` contains a peak matching the
+#'     first peak (row) of `y` or `NA_real_` if the peak does not match.
+#'
+#' @param ... ignored
+#'
+#' @author Johannes Rainer, Michael Witting, based on the code from the ref
+#'
+#' @importFrom clue solve_LSAP
+#'
+#' @importFrom stats complete.cases
+#'
+#' @references
+#'
+#' Xing S, Hu Y, Yin Z, Liu M, Tang X, Fang M, Huan T. Retrieving and Utilizing
+#' Hypothetical Neutral Losses from Tandem Mass Spectra for Spectral Similarity
+#' Analysis and Unknown Metabolite Annotation. *Anal Chem.*
+#' 2020 Nov 3;92(21):14476-14483. doi: 10.1021/acs.analchem.0c02521.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' ## Define spectra
+#' x <- cbind(mz = c(10, 36, 63, 91, 93), intensity = c(14, 15, 999, 650, 1))
+#' y <- cbind(mz = c(10, 12, 50, 63, 105), intensity = c(35, 5, 16, 999, 450))
+#' ## The precursor m/z
+#' pmz_x <- 91
+#' pmz_y <- 105
+#'
+#' ## joinPeaksGnps finds 4 matches
+#' map <- joinPeaksGnps(x, y, pmz_x, pmz_y)
+#' map
+#'
+#' ## Calculate similarity
+#' gnps(map$x, map$y)
+gnps <- function(x, y, ...) {
```

Comment:
Should this function not rather be added to `MsCoreUtils`? It doesn't use or return `Spectra` objects.

Reply:
I originally also thought the same. But this function needs to be called in combination with `joinPeaksGnps` to get GNPS scores. Putting it into `MsCoreUtils` I'm afraid people might get the impression that this alone calculates GNPS similarity. But I have no strong opinion on that. I could also add it to `MsCoreUtils` if you think that's better.

Reply:
Does that mean that the function on its own is likely to be useless? If so, them I would suggest to keep it in `Spectra` but not export it. If there is a use for it, then we a move to `MsCoreUtils` could be considered.

Reply:
Maybe I should explain  better: in order to calculate GNPS scores one would have to call:

```r
compareSpectra(sps, sps2, MAPFUN = joinPeaksGnps, FUN = gnps)
```

so we need to export `gnps`.  I could move `gnps` to `MsCoreUtils`, but then I would need also to add a function `join_gnps` there (that does the peak matching as in GNPS) and mention in the documentation that `gnps` only returns GNPS scores, if the peaks were first matched with `join_gnps`.

Reply:
OK, in the end I agree. I will add the `join_gnps` and `gnps` functions to the `MsCoreUtils` package and then update the PR here. So you will get another review request today over in `MsCoreUtils` :)